[Nicholas Tan] iP

NicholasTYD opened the pull request:
# DukePRO
- Its Duke.
- But **PRO**
- Also iP compliant!

## Reasons why you should use merge this PR
1. Why not
2. :grinning:

```java
public class Main {
    public static void main(String[] args) {
        System.out.println("A fenced code block (with syntax highlighting)");
    }
}
```

> That’s a nice argument Senator, why don’t you back it up with a source? - The [Source](https://www.blank.org/)

## Features
- [x] Managing Tasks
- [x] Managing deadlines
- [ ] Reminders (Coming never)
- [X] Contains a `main` method

Commit:
Add Gradle support

NicholasTYD committed:
Completed level 1

NicholasTYD committed:
Completed level 2

NicholasTYD committed:
renamed storeTask to addTask

NicholasTYD committed:
Completed lvl 3

NicholasTYD committed:
Completed level 4

NicholasTYD committed:
Added testing, and changed output grammar slightly in duke

NicholasTYD committed:
Completed lvl 5

NicholasTYD committed:
Completed lvl 6

NicholasTYD committed:
Reformatted switch case

NicholasTYD committed:
Make Duke non static

NicholasTYD committed:
Add SaveSystem

NicholasTYD committed:
Add test cases for saving

NicholasTYD committed:
Add time and date for deadline

NicholasTYD committed:
Add Date and Times for event and deadlines

NicholasTYD committed:
Merge branch 'branch-Level-7'

NicholasTYD committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

NicholasTYD committed:
Fix merge bugs

NicholasTYD committed:
Fix save bug caused by "|" input

NicholasTYD committed:
Add and optimise Storage class

NicholasTYD committed:
Extract code from Duke to Parser, TaskList, UI classes

NicholasTYD committed:
Implement Parser class and optimise code

NicholasTYD committed:
Add all classes to package duke

NicholasTYD committed:
Revert "Add all classes to package duke"

This reverts commit bbca9e75a70b400a4ac00b0020b741ca97e71dc3.

NicholasTYD committed:
Package all files into package duke

NicholasTYD committed:
Merge remote-tracking branch 'origin/add-gradle-support'

NicholasTYD committed:
Add JUnit tests

NicholasTYD committed:
Modify build.gradle application mainClassName

NicholasTYD committed:
Setup checkstyle

NicholasTYD committed:
Add JavaDoc comments

NicholasTYD committed:
Tweak code to comply with coding standard.

NicholasTYD committed:
Add ability to find tasks

NicholasTYD committed:
Add staged files missed by previous commit

NicholasTYD committed:
Merge branch 'branch-A-JavaDoc'

NicholasTYD committed:
Merge branch 'branch-A-CodingStandard'

NicholasTYD committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/TaskList.java

NicholasTYD committed:
Fix checkstyle violations

NicholasTYD committed:
Add javaFx

NicholasTYD committed:
Refine GUI

NicholasTYD committed:
Fix some checkstyle violations

NicholasTYD committed:
Re-add border to Duke chatbot

NicholasTYD committed:
Shift welcome message into Ui class

NicholasTYD committed:
Add javadocs for constructors in Event, Deadline and Todo

NicholasTYD committed:
Change access modifier of Task constructor from public to protected

NicholasTYD committed:
Merge branch 'branch-Level-10'

amresh2000 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,130 @@
+        return ui.getExitMessage();
+    }
+
+    private void closeApplicationAfterDelay() {
```
Good naming convention of methods throughout the codebase

amresh2000 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,83 @@
+        default:
+            String[] str = input.split(" ", 2);
+            String cmd = str[0];
+            if (str.length == 1
```
Good job following all indentation code styles throughout code

amresh2000 reviewed (commented):
Overall I found your code easy to read. Followed all of the java coding style conventions. Job well done 😃

Berted commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,14 @@
+/**
+ * Encapsulates custom Duke related exceptions thrown when using duke.
+ */
+public class DukeException extends RuntimeException {
```
I like how you created separated each type of DukeException into its own class to increase the clarity of the type of DukeException that occurred in the code.

Berted commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,130 @@
+        return tasks.size();
+    }
+
+    public void loadTask(Task task) {
```
Currently, `loadTask` gives an impression that it is loading from a string or file. Maybe the private-scoped `addTask` can just be made public instead?

Berted commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,177 @@
+        return getMessage(enumerateTasks(tasks, reply));
+    }
+
+    private String enumerateTasks(ArrayList<Task> tasks, String str) {
```
I like how you separated the logic for enumerating tasks into its own method to improve code reuse!

Berted commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,83 @@
+            }
+            parsedOutput[0] = cmd;
+
+            switch (cmd) {
```
Perhaps parsing the command can be done prior to the first `switch` statement to avoid [deep nesting](https://nus-cs2103-ay2223s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-deep-nesting)?

Berted commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,83 @@
+            case "find":
+            case "mark":
+            case "unmark":
+            case "delete":
```
Maybe add `// Fallthrough` comment to [explicitly state its intended behaviour](https://se-education.org/guides/conventions/java/index.html#layout)?

Berted reviewed (commented):
LGTM! Overall the code is very cleanly written and I learned some stuff that I could apply to my iP.

NicholasTYD committed:
change MainClassName in build.gradle to Launcher

NicholasTYD committed:
Add assertions to Duke and Task class

There are no assertions in Duke and Task classes.

There are some assumptions present that the code in said classes
use while running (E.g getResponse() in Duke assumes a non-null
UI reference in its class).

Adding assert statements ensure those references are non-null and
truly present which can help in debugging code.

Let's add assert statements to check that references are non null
in Duke and Task classes.

Using assertions is preferable over if statements as it conveys
that the code is used for debugging purposes and makes the code
look cleaner.

NicholasTYD committed:
Improve code quality in UI class.

There are several methods in UI class that contains magic strings.

Magic strings is not good for extensibility as it is harder for
people to understand its purpose.

Lets use named constants to denote such strings and return them.

This way it will be more easily understood by people reading the
code.

NicholasTYD committed:
Remove redundant method loadTask in class TaskList

NicholasTYD committed:
Refactor storage class using SLAP

NicholasTYD committed:
Refactor Parser according to SLAP

NicholasTYD committed:
Improve Parser class code quality.

parseInput() method is very long with deep nesting. It also
has many level of abstraction, violating the SLAP principle.
It is difficult to understand the code. There are also magic
literals present.

Let's apply the SLAP principle and break the parseInput() method
into multiple smaller methods, as well as refactor code to reduce
the level of nesting.

It would be easier to read and understand the code in the future,
as well as making it easier to debug.

NicholasTYD committed:
Fix ParserTest test cases

NicholasTYD committed:
Improve code quality of Duke class through SLAP and name constants.

NicholasTYD committed:
Merge pull request #1 from NicholasTYD/branch-A-Assertions

Add assertions to Duke and Task class

NicholasTYD committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/duke/Duke.java

NicholasTYD committed:
Add friendlier syntax

Commands are case sensitive, and there are no shortcuts.

Case insensitive commands and shortcuts for commands are more user
friendly as users might make not remember the correct cases when
typing in commands. Shortcuts will also help expert users navigate
the app faster.

Lets,
- Update commands so that they become case insensitive.
- Implement some short forms for commands so that it is faster for
expert users to use.

NicholasTYD committed:
Merge pull request #3 from NicholasTYD/branch-BCD-Extension

Add friendlier syntax

NicholasTYD committed:
Fix checkstyle errors

NicholasTYD committed:
Add error handling for invalid mark/unmark/delete number input.

NicholasTYD committed:
Add help command

NicholasTYD committed:
Modify dialog box to fit more elements

NicholasTYD committed:
Add auto resizable dialog box

NicholasTYD committed:
Improve GUI look

NicholasTYD committed:
Improve compactness of dialogue box

NicholasTYD committed:
Extend Duke speech border

NicholasTYD committed:
Add javadoc for getHelpMessage method in UI class

NicholasTYD committed:
Update Dialog box to scale with text

NicholasTYD committed:
Fix dialog box bugs

NicholasTYD committed:
Fix checkstyle error

NicholasTYD committed:
Update README.md

NicholasTYD committed:
Update README.md

NicholasTYD committed:
Fix command parser not recognizing caps of short form commands

NicholasTYD committed:
Update README.md

NicholasTYD committed:
Update README.md

NicholasTYD committed:
Add assert statement when loading tasks in Storage class

NicholasTYD committed:
Merge branch 'branch-temp-sync'

NicholasTYD committed:
Make find command case insensitive

NicholasTYD committed:
Update README.md

NicholasTYD committed:
Update README.md

NicholasTYD committed:
Update README.md

NicholasTYD committed:
Update README.md

NicholasTYD committed:
Update README.md

NicholasTYD committed:
Update README.md

NicholasTYD committed:
Fix magic literal in Deadline, Event, Main, Task, Todo Classes

NicholasTYD committed:
Add Ui representative screenshot

NicholasTYD committed:
Fix find bug not being case insensitive when search word is in caps.

NicholasTYD committed:
Modify save location to hard disk

NicholasTYD committed:
Update README.md
